### DISCUSSION TITLE
arithmetic exponentiation operator proposal

### DISCUSSION THREAD

**Discussion by bindsdev** [carbon-language/carbon-lang#1835] [Language design]
I propose that syntax and functionality for exponentiation is implemented. Currently, if you want to raise some number `x` to a power of 2, you must do `x * x`. This is fine if you only want to raise `x` to itself *once*. But, if there is ever a case where you want to raise `x` to a power greater than 2, it could get ugly quickly. The operator would be like Python's exponentiation operator, which is `**`. Operator overloading for this could be implemented using `Raise` and `RaiseWith(U)` or `Exponent` and `ExponentWith(U)` interfaces.

With this operator, raising `x` to a power of 3 could look like `x ** 3` instead of `x * x * x`.

**Comment by nigeltao** [comment]
Be careful, though, that `x**y` isn't ambiguous with the existing `x * *y` syntax where the first star is multiplication (a binary operator) and the second star is a pointer dereference (a unary operator).

Related: the `deref p` syntax idea from #1454 (which is admittedly different from C++ syntax).

**Comment by bindsdev** [comment]
> Be careful, though, that `x**y` isn't ambiguous with the existing `x * *y` syntax where the first star is multiplication (a binary operator) and the second star is a pointer dereference (a unary operator).
> 
> Related: the `deref p` syntax idea from #1454 (which is admittedly different from C++ syntax).

Indeed. With this concern, the operator could be potentially modified to `^+`. I do not see any ambiguity with this as `+` itself is not an existing unary operator from my knowledge. My first thought was `^*` but `x ^ *y` (XOR and dereference) could become ambiguous with `x^*y`.

**Comment by OlaFosheimGrostad** [comment]
`x^^n`

**Comment by tkoeppe** [comment]
Does this really need to be an _operator_, rather than, say, a standard library function? Is this operation common enough to warrant a dedicated syntax (and requiring everyone to learn that syntax)?

**Comment by bindsdev** [comment]
> Does this really need to be an _operator_, rather than, say, a standard library function? Is this operation common enough to warrant a dedicated syntax (and requiring everyone to learn that syntax)?

This is a fair question. But personally, I would much rather have an operator than a function call, mainly to not have to append `.pow` or `.raise` to every number I want to exponentiate. If unnecessary syntax is a concern though, I suppose a function would be fine

**Comment by chandlerc** [comment]
FYI, moving from an issue to a discussion.

**Comment by YagaoDirac** [comment]
This should be done with a math std library. It shouldn't be done as language feature. Carbon as a container language shares the same philosophy with c++. 
But also, it can be done like:
{scope.enable_rich_arith_operator(true);
var x:int = 42;
x**2;//OK
}
